### Step 1: Import Necessary Libraries

In this step, we import the essential libraries required for data validation:
- `pandas` for data manipulation.
- `os` and `json` for handling file paths and configurations.
- `sys` to check the running environment (Jupyter notebook or script).


In [1]:
import pandas as pd # type: ignore
import os
import json
import sys


### Step 2: Load Configuration and Data Paths

Here, we load the configuration file from the main project directory. We use a dynamic approach to detect whether the code is running in a Jupyter notebook or as a script:

- **Jupyter Notebook Detection**: The code checks if it's running in a Jupyter notebook environment by detecting the presence of `ipykernel` in the system modules.
- **Project Root Directory**: Based on the environment, it correctly sets the `project_root` directory. If it's running in a Jupyter notebook, it assumes the notebook is in the `notebooks` directory and adjusts the path accordingly.


In [2]:
import os
import json

# Set the project root directory directly to the Customer-Churn-Analysis folder
project_root = os.path.abspath(os.path.join(os.getcwd(), '..', '..'))

# Load configuration file from the main directory
config_path = os.path.join(project_root, 'config.json')
if not os.path.exists(config_path):
    raise FileNotFoundError(f"Config file not found at {config_path}")

with open(config_path, 'r') as f:
    config = json.load(f)

# Convert relative paths in the configuration file to absolute paths
raw_data_path = os.path.join(project_root, config['raw_data_path'])

print(f"Raw data path: {raw_data_path}")


Raw data path: /Volumes/Bhavesh/Customer-Churn-Analysis/data/raw/Dataset (ATS)-1.csv


### Step 3: Load the Raw Data

In this step, we load the raw dataset from the path specified in the configuration file. This dataset represents the unprocessed data that will undergo validation checks.


In [3]:
# Step 3: Load the raw data
df = pd.read_csv(raw_data_path)
print(f"Original dataset loaded from {raw_data_path}. Shape: {df.shape}")
df.head()


Original dataset loaded from /Volumes/Bhavesh/Customer-Churn-Analysis/data/raw/Dataset (ATS)-1.csv. Shape: (7043, 10)


gender  SeniorCitizen Dependents  tenure PhoneService MultipleLines  \
0  Female              0         No       1           No            No   
1    Male              0         No      34          Yes            No   
2    Male              0         No       2          Yes            No   
3    Male              0         No      45           No            No   
4  Female              0         No       2          Yes            No   

  InternetService        Contract  MonthlyCharges Churn  
0             DSL  Month-to-month           29.85    No  
1             DSL        One year           56.95    No  
2             DSL  Month-to-month           53.85   Yes  
3             DSL        One year           42.30    No  
4     Fiber optic  Month-to-month           70.70   Yes

### Step 4: Perform Data Integrity and Validity Checks

This step defines several functions to perform various data validation checks:

- **`check_missing_values(df)`**: Identifies and reports any missing values in the dataset.
- **`check_duplicates(df)`**: Detects duplicate rows but does not remove them.
- **`check_data_types(df, expected_types)`**: Validates that each column has the expected data type.
- **`check_value_ranges(df, value_ranges)`**: Ensures that numerical values fall within defined ranges.
- **`check_unique_columns(df, unique_columns)`**: Checks for unique values in specified columns, such as `customerID`.
- **`check_consistency(df)`**: Verifies the consistency of derived columns, like ensuring `TotalCharges` equals `tenure * MonthlyCharges`. This check is skipped if `TotalCharges` is not present, which is expected in the raw data.

These functions collectively ensure that the dataset is valid and ready for analysis.


In [4]:
def check_missing_values(df):
    missing_values = df.isnull().sum()
    if missing_values.any():
        print("Missing values found:")
        print(missing_values[missing_values > 0])
    else:
        print("No missing values found.")
    return missing_values

def check_duplicates(df):
    duplicate_rows = df[df.duplicated()]
    if not duplicate_rows.empty:
        print(f"Found {len(duplicate_rows)} duplicate rows:")
        print(duplicate_rows)
    else:
        print("No duplicate rows found.")
    return duplicate_rows

def check_data_types(df, expected_types):
    mismatched_types = {}
    for column, expected_type in expected_types.items():
        if column in df.columns and df[column].dtype != expected_type:
            mismatched_types[column] = df[column].dtype
    if mismatched_types:
        print("Data type mismatches found:")
        print(mismatched_types)
    else:
        print("All data types are as expected.")
    return mismatched_types

def check_value_ranges(df, value_ranges):
    out_of_range = {}
    for column, (min_value, max_value) in value_ranges.items():
        if column in df.columns:
            out_of_bounds = df[(df[column] < min_value) | (df[column] > max_value)]
            if not out_of_bounds.empty:
                out_of_range[column] = out_of_bounds
    if out_of_range:
        print("Out of range values found:")
        for column, values in out_of_range.items():
            print(f"{column}:")
            print(values)
    else:
        print("All values are within the expected range.")
    return out_of_range

def check_unique_columns(df, unique_columns):
    not_unique = {}
    for column in unique_columns:
        if column in df.columns and df[column].duplicated().any():
            not_unique[column] = df[column][df[column].duplicated()]
    if not_unique:
        print("Columns with non-unique values found:")
        for column, values in not_unique.items():
            print(f"{column}:")
            print(values)
    else:
        print("All specified columns have unique values.")
    return not_unique

def check_consistency(df):
    if 'TotalCharges' in df.columns and 'tenure' in df.columns and 'MonthlyCharges' in df.columns:
        inconsistent = df[df['TotalCharges'] != df['tenure'] * df['MonthlyCharges']]
        if not inconsistent.empty:
            print("Inconsistent values found:")
            print(inconsistent[['tenure', 'MonthlyCharges', 'TotalCharges']])
        else:
            print("All values are consistent.")
    else:
        print("Skipping consistency check for 'TotalCharges' because it is not present.")


### Step 5: Execute the Data Validation

Now, we execute the data validation functions defined earlier on our dataset. This will ensure the dataset meets the required integrity and validity standards, identifying any issues that need to be addressed before further analysis.

The checks performed include:
- **Missing values detection**
- **Duplicate row identification**
- **Data type verification**
- **Value range validation**
- **Uniqueness of specific columns**
- **Consistency of calculated values**

Since `TotalCharges` is not part of the raw dataset, the consistency check related to this column is skipped.


In [5]:
# Execute the data validation checks
check_missing_values(df)
check_duplicates(df)
expected_types = {
    'customerID': 'object',
    'gender': 'object',
    'SeniorCitizen': 'int64',
    'Partner': 'object',
    'Dependents': 'object',
    'tenure': 'int64',
    'PhoneService': 'object',
    'MultipleLines': 'object',
    'InternetService': 'object',
    'OnlineSecurity': 'object',
    'OnlineBackup': 'object',
    'DeviceProtection': 'object',
    'TechSupport': 'object',
    'StreamingTV': 'object',
    'StreamingMovies': 'object',
    'Contract': 'object',
    'PaperlessBilling': 'object',
    'PaymentMethod': 'object',
    'MonthlyCharges': 'float64',
    'TotalCharges': 'float64',
    'Churn': 'object'
}
check_data_types(df, expected_types)
value_ranges = {
    'tenure': (0, 100),
    'MonthlyCharges': (0, 150),
    'TotalCharges': (0, 10000)
}
check_value_ranges(df, value_ranges)
unique_columns = ['customerID']
check_unique_columns(df, unique_columns)
check_consistency(df)


No missing values found.
Found 103 duplicate rows:
      gender  SeniorCitizen Dependents  tenure PhoneService MultipleLines  \
100     Male              0         No       1          Yes            No   
772   Female              0         No       1          Yes            No   
885     Male              0         No       1          Yes            No   
964     Male              0         No       1          Yes            No   
987     Male              0         No       1          Yes            No   
...      ...            ...        ...     ...          ...           ...   
6854    Male              0         No       2          Yes            No   
6875  Female              0         No      64          Yes            No   
6924    Male              0         No       1          Yes            No   
6977    Male              0         No      24          Yes            No   
7010  Female              1         No       1          Yes           Yes   

     InternetService    

### Summary of Results

In this notebook, we performed several key data integrity and validity checks on our raw dataset. The following outcomes were observed:

1. **Missing Values Check**:
   - **Result**: No missing values were found in the dataset, ensuring that all records are complete and can be used for further analysis without the need for imputation or removal of records.

2. **Duplicate Rows Check**:
   - **Result**: We identified 103 duplicate rows within the dataset. These rows have identical values across all columns, suggesting that they might be repeated entries.

3. **Data Type Validation**:
   - **Result**: All columns were found to have the expected data types, meaning the dataset's structure is consistent with our expectations and ready for processing.

4. **Value Range Validation**:
   - **Result**: All numerical columns fall within the expected ranges, indicating that there are no outlier values that could skew the analysis.

5. **Unique Column Check**:
   - **Result**: All specified columns (e.g., `customerID`) were found to contain unique values, confirming that each record represents a distinct entity in the dataset.

6. **Consistency Check**:
   - **Result**: The consistency check for `TotalCharges` was skipped since this column was derived during feature engineering and is not present in the raw dataset.

### Explanation: Handling of Duplicate Rows

During the data validation process, we encountered 103 duplicate rows in the dataset. After careful consideration, we decided **not to remove these duplicates**. Here’s the rationale behind this decision:

1. **Data Integrity and Completeness**:
   - Removing duplicates might inadvertently remove important customer data, potentially leading to loss of valuable information. For example, if two identical records exist, both may represent legitimate transactions or entries that should be retained in the analysis.

2. **Impact on Downstream Analysis**:
   - The presence of duplicate rows does not significantly affect our downstream analysis, such as clustering or predictive modeling, especially when dealing with large datasets. Additionally, removing duplicates could disrupt the continuity of our data processing pipeline, requiring further adjustments to the analysis scripts.

3. **Project Requirements**:
   - Our primary goal is to ensure the data's validity and integrity, not necessarily to cleanse it of all duplicates unless they pose a direct risk to the analysis. Since the duplicates identified do not significantly impact the overall analysis, we have chosen to retain them.

This decision allows us to maintain the dataset's completeness while ensuring that the integrity and consistency checks are thorough and aligned with our project's objectives.
